### PR TITLE
build: Use a zip instead of dmg for macOS releases

### PR DESCRIPTION
**Pull request by fanquake**
Implements #18128 (mostly). From discussion on that issue we've got 3 `ACK`s (to move from .dmg to .zip), 1 `~0` and 1 `NACK`. So I've opened this PR to give a better overview of what the changes might actually look like. 

Gitian builds currently work, however I haven't tested the signer, and that likely needs some additional changes. As mentioned we can drop quite a few packages from depends. There's some commits here that can cleaned up, squashed or dropped.

If you're compiling macOS, `make appbundle` will (as before) build a dynamically linked `Bitcoin Core.app`. `make deploy` will still build a statically linked `dist/Bitcoin Core.app` and will now also create a `dist/Bitcoin-Core.zip`.

Compiling on Linux there has never been a `make appbundle` rule, and with these changes, `make deploy` now behaves the same as on macOS. The `make deploydir` rule has  been removed. 

Initially I  was unconvinced (of a swap from .dmg to .zip), however I have come around after looking at these changes. If there is not agreement that "this is the way", then I'll revive #18151 and fix it up to include `xorriso` in depends etc.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Gitian Build seems to work: https://bitcoin.jonasschnelli.ch/gitian/build/142 
Maybe the ZIP filename could still contain the phrase "macOS" or "osx".

Conceptual, I'm currently ~0 (I [commented](https://github.com/bitcoin/bitcoin/issues/18128#issuecomment-585323098) on the concept before).
I think it is nice to simplify the build system (probably makes it much easier for guix). Though, I'm currently unsure if this slightly worsens the UX for macOS users. If it does, should we reduce the UX to simplify the build-system (is that an acceptable tradeoff)?

However, if there is a difference in user experience between a zip and a dmg, it is probably marginal.

**Comment by MarcoFalke** [comment]
Is there any "official" mac documention or user surveys that suggest which one is the recommended or preferred way?

**Comment by dongcarl** [comment]
Concept ACK! Great to see the amount of code we can remove.

-----

> (probably makes it much easier for guix).

Just to make myself clear, I am _**for**_ this change because it is a great simplification for a marginal difference in user experience and not because of anything specific to Guix (my current implementation of the Guix macOS build is already able to build the `dmg`). :relaxed:

**Comment by Sjors** [comment]
You can add a symlink to `/Applications` to the archive and call it "Drag Here".
Schermafbeelding 2020-05-27 om 18 29 17

With that it's close enough to what macOS users are used to, with a nice reduction in dependencies and complexity.

You could also keep the `.DS_Store` and `.background` folder around for full branding and a nicer layout, but that's pretty marginal.

**Comment by fanquake** [comment]
> You could also keep the .DS_Store and .background folder around for full branding and a nicer layout, but that's pretty marginal.

"Keeping that around" would mostly defeat the point of these changes.

**Comment by Sjors** [comment]
> "Keeping that around" would mostly defeat the point of these changes.

It still gets rid of `genisoimage` and `libdmg-hfsplus`, but indeed `.DS_Store` and `.background` require a lof of tools themselves. Just the symlink should get us close enough to the original UX.

**Comment by fanquake** [comment]
Pushed a few more removals, updated the descriptors and fixed whitespace.

> Is there any "official" mac documention or user surveys that suggest which one is the recommended or preferred way?

The Apple documentation is mainly focused on using Xcode and App Store distribution: https://help.apple.com/xcode/mac/current/#/devac02c5ab8. There is a section on distribution outside the App Store: https://help.apple.com/xcode/mac/current/#/dev033e997ca, however I couldn't quickly find any mention of using `.dmg`s. There are some instructions on "macOS installer packages" (`.pkg`): https://help.apple.com/xcode/mac/current/#/deve51ce7c3d.

There seem to be at least some companies that are using a combination of `.dmg` and `.zip` for distribution. i.e: [Sublime Merge](https://www.sublimemerge.com/) is distributed as a `.zip`, while [Sublime Text](https://www.sublimetext.com/) is a `.dmg`.

> Just the symlink should get us close enough to the original UX.

I'm not aware of any projects that are using a symlink inside a `.zip` approach. The Bitcoin Core GUI is mostly [designed for power users of bitcoin](https://github.com/bitcoin/bitcoin/issues/17395#issuecomment-550339984), and I think the intersection of power users, and people who don't know how to handle a `.zip` file would be very small. 

Also, as far as I'm aware, you can open and use an app bundle from any directory on macOS, and it makes  little to no difference to how the app actually runs?

**Comment by Sjors** [comment]
> I'm not aware of any projects that are using a symlink inside a .zip approach

That's because most projects use a DMG with a symlink and background image. We have a good reason to avoid that, namely to make deterministic builds easier. Even as an experienced macOS user, I find the symlink useful; it saves me from opening a new Finder window, navigating to Applications and then dragging between two windows. And it's a trivial thing to add.

> Also, as far as I'm aware, you can open and use an app bundle from any directory on macOS, and it makes little to no difference to how the app actually runs?

Whenever I deal with applications that don't have an installer or a symlink, I very often make the mistake of running it from the Download folder. That works, until I accidentally clean up the download folder. There's no obvious alternative to drag the program to.

File: Makefile.am [hunk 1]
```diff
@@ -116,50 +111,18 @@ osx_volname:
 	echo $(OSX_VOLNAME) >$@
 
 if BUILD_DARWIN
-$(OSX_DMG): $(OSX_APP_BUILT) $(OSX_PACKAGING) $(OSX_BACKGROUND_IMAGE)
```
Comment (fanquake) [inline Makefile.am:119]: One possible simplification after this change, would be to make the `otool` and `installnametool` checks no longer cross-compiling dependent. Then we should be able to have a single `OSX_ZIP` rule, and drop the need for `BUILD_DARWIN`.

**Comment by luke-jr** [comment]
Can we add the ZIP and remove the DMG in two steps?

IMO it'd be nice to build both for 0.21, but only distribute the ZIP. That way if it causes too much problems/confusion we can just change it back without rebuilds/etc.

**Comment by fanquake** [comment]
I might come back to this in future, but closing for now as I've made some changes to the existing macdeploy process. Any discussion should continue in #18128.

**Pull request closed without merging**